<a href="https://colab.research.google.com/github/SRIDEV93/Python-for-DataAnalysis/blob/main/LLoyds_Stock_Market_Analysis_Since_Pandemic.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#LLoyds Stock Market Analysis
Stock Market Analysis is one of the best usecase interms of Data Science in Finance. So, I want to Analyze the LLoyds stock Market using Python.I used the yfinance API of Yahoo Finance for collecting the stock price data.

**LLoyds Stock Market Dataset Since Pandemic**

Here I am downloading dataset from yfinance. yfinance is a Yahoo Finance, which is one of the most popular websites to collect the stock price data. I am interested to see the stock analysis of the LLoyds bank since Pandemic to till date. To Analyze this data everytime, I have to use the yfinace API, which is provided by Yahoo Finance to collect the latest stock price data. The Yahoo Finance API is installed using the pip command.

In [53]:
!pip install yfinance

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


# Libraries required

In [54]:
import pandas as pd
import yfinance as yf
import datetime 
from datetime import timedelta, date
import plotly.graph_objects as go
import plotly.express as px

In [55]:
# d1 is today's date
d1 = date.today().strftime("%Y-%m-%d")
end_date = d1

#d2 is the start date of the pandemic in UK
d2 = datetime.date(2020,1,31)
start_date = d2.strftime("%Y-%m-%d")

# LLOY.L is the LLoyds banking group in London, United Kingdom

# I don't want to see the downloaded progress, So I put progress= False
data = yf.download('LLOY.L', start=start_date, end=end_date, progress=False)

#First five rows of the data
print(data.head())


                 Open       High        Low      Close  Adj Close     Volume
Date                                                                        
2020-01-31  58.480000  58.480000  56.349998  56.790001  48.708920  179297410
2020-02-03  56.980000  57.200001  56.189999  56.549999  48.503075  153681928
2020-02-04  57.320000  57.590000  56.830002  57.500000  49.317890  113644729
2020-02-05  57.660000  58.327999  57.139999  57.139999  49.009117  194283019
2020-02-06  57.650002  58.270000  56.845001  56.950001  48.846153  142821635


In [56]:
shape = data.shape
shape

(641, 6)

In [57]:
for columns in data.columns:
  print(columns)

Open
High
Low
Close
Adj Close
Volume


Here the shape of the dataframe looks like, it has 7 columns. But it is not true.

The shape of the dataframe is (641,6) i.e., 641 rows and 6 columns.

Here, Date is not a column, it is the index of this dataset. So we have to convert this index into column.



# Converting Date(Index) into Column

In [58]:
data["Date"] = data.index
data = data[["Date", "Open", "High", "Low", "Close", "Adj Close", "Volume"]]

data.reset_index(drop=True, inplace = True)

##drop = True
#If you set drop = True, reset_index will delete the index instead of inserting it back into columns of the df.
#If you set drop = True, the current index will be deleted entirely and the numeric index will replace it.

## inplace = True
# When inplace = True, the data is updated inplace, which means it will return nothing, and df is now updated.

print(data.head())

        Date       Open       High        Low      Close  Adj Close     Volume
0 2020-01-31  58.480000  58.480000  56.349998  56.790001  48.708920  179297410
1 2020-02-03  56.980000  57.200001  56.189999  56.549999  48.503075  153681928
2 2020-02-04  57.320000  57.590000  56.830002  57.500000  49.317890  113644729
3 2020-02-05  57.660000  58.327999  57.139999  57.139999  49.009117  194283019
4 2020-02-06  57.650002  58.270000  56.845001  56.950001  48.846153  142821635


In [59]:
shape = data.shape
shape

(641, 7)

In [60]:
for column in data.columns:
  print(column)

Date
Open
High
Low
Close
Adj Close
Volume


Now date is added as an additional column. That's why the shape of the dataframe is (641,7) 

The candlestick chart is the best place to start whenever you are analyzing the stock market. The candlestick chart can be used to analyze movements in stock prices. Here's how you can visualize the Lloyds stock price chart using candlestick chart.

In [61]:
figure = go.Figure(data=[go.Candlestick(x=data["Date"], open=data["Open"], high=data["High"], low=data["Low"], close=data["Close"])])

figure.update_layout(title = "LLoyds Stock Price Analysis", xaxis_rangeslider_visible=False)

figure.show()

Here is how to visualize LLoyds's stock price using a bar plot to analyze the stock market, especially in the long run.

In [62]:
figure = px.bar(data, x = "Date", y="Close")
figure.show()

The range slider is a useful tool for analyzing the stock market, allowing you to choose a time period and analyze it. Here's how to add one:

In [63]:
figure = px.line(data, x = "Date", y='Close', title = 'LLoyds Stock Market Analysis with Rangeslider')
figure.update_xaxes(rangeslider_visible=True) 
figure.show()

Another interative feature that we can add for the stock market analysis is time period selectors. Time period selectors are like buttons that show you the graph of a specific time period. Here is how we cn add buttons for selecting the time period for stock market analysis.

In [64]:
figure = px.line(data, x="Date", y="Close", title='Lloyds Stock Market Analysis with Time period selectors')

figure.update_xaxes(
    rangeselector=dict(
        buttons=list([
            dict(count=3, label='3m', step="month", stepmode="backward"),
            dict(count=6, label='6m', step="month", stepmode="backward"),
            dict(count=1, label='1y', step="year", stepmode="backward"),
            dict(count=1.5, label='1.5y', step='year', stepmode='backward'),
            dict(count=2, label='2y', step='year', stepmode='backward'),
            dict(step="all")
  
        ])
    )
)
figure.show()

In [65]:
figure = go.Figure(data=[go.Candlestick(x=data["Date"], open=data["Open"], high=data["High"], low=data["Low"], close=data["Close"])])

figure.update_layout(title = "LLoyds Stock Price Analysis", xaxis_rangeslider_visible=False)

figure.update_xaxes(
    rangeselector=dict(
        buttons=list([
            dict(count=1, label='1d', step='day',stepmode = 'backward'),
            dict(count=1, label='1m', step='month', stepmode ='backward'),
            dict(count=3, label='3m', step='month', stepmode='backward'),
            dict(count=6, label='6m', step='month', stepmode='backward'),
            dict(count=1, label='1y', step='year', stepmode='backward'),
            dict(count=1.5, label='1.5y', step='year', stepmode='backward'),
            dict(count=2, label='2y', step='year', stepmode='backward'),
            dict(step="all")

        ])
    )
)



figure.show()

The weekend or holiday season always affects the stock market. So if you want to remove all the records of the weekend trends from your stock market visualization below is how you can do it

In [66]:
figure = px. scatter(data, x ='Date', y = 'Close', range_x=['2020-01-31','2022-08-17'], title = 'LLoyds Stock Market Analysis by Hiding Weekend Gaps')

figure.update_xaxes(
    rangebreaks = [
        dict(bounds=["sat","sun"])
    ]
)
figure.show()